File: app/src/main/java/ru/tsu/quizium/QuizListActivity.java
package ru.tsu.quizium;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.GenericTypeIndicator;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

public class QuizListActivity extends AppCompatActivity {

    private final ArrayList<Quiz> quizList = new ArrayList<Quiz>();
    private final ArrayList<Category> categoryList = new ArrayList<Category>();
    private final ArrayList<RecyclerViewItem> resultItemList = new ArrayList<RecyclerViewItem>();

    private FirebaseAuth mAuth;
    private DatabaseReference dbRef;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_quiz_list);

        dbRef = FirebaseDatabase.getInstance().getReference();


        InitCategories();
        InitQuizData();
        InitRecyclerViewDataList();
        RecyclerView recyclerView = findViewById(R.id.quizRecyclerView);
        RecyclerViewAdapter adapter = new RecyclerViewAdapter(this, resultItemList);
        recyclerView.setAdapter(adapter);
    }

    private void InitRecyclerViewDataList(){
        for(int i = 0; i < 4; i++){
            resultItemList.add(new RecyclerViewItem(categoryList.get(i)));
            for(int j = 0; j < 4; j++){
                resultItemList.add(new RecyclerViewItem(quizList.get(j)));
            }
        }
    }

    private void InitQuizData() {
        for(int i = 0; i < 4; i++){
            Category category = categoryList.get(i);
            quizList.add(new Quiz("Quiz 1", "Description of Quiz 1 Description of Quiz 1 Description of Quiz 1 Description of Quiz 1 Description of Quiz 1 Description of Quiz 1 Description of Quiz 1", "admin", R.drawable.english_grammar, category.getName()));
            quizList.add(new Quiz("Quiz 2", "Description of Quiz 2", "admin", R.drawable.english_vocabulary, category.getName()));
            quizList.add(new Quiz("Quiz 3", "Description of Quiz 3", "admin", R.drawable.math, category.getName()));
            quizList.add(new Quiz("Quiz 4", "Description of Quiz 4", "admin", R.drawable.math, category.getName()));
        }
    }

    private void InitCategories(){
        for(int i = 0; i < 4; i++){
            categoryList.add(new Category("Category " + i));
        }
    }
}

File: app/src/main/java/ru/tsu/quizium/QuizListActivity.kt
package ru.tsu.quizium

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.database.*
import ru.tsu.quizium.dto.Category
import ru.tsu.quizium.dto.Quiz
import ru.tsu.quizium.dto.RecyclerViewItem
import java.util.*

class QuizListActivity : AppCompatActivity() {

    private val categoryList = mutableListOf<Category>()
    private val resultItemList = ArrayList<RecyclerViewItem>()

    private val database by lazy { FirebaseDatabase.getInstance() }
    private val themeRef by lazy { database.getReference("theme") }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_quiz_list)

        getQuizzes()
    }

    private fun getQuizzes(){
        themeRef.addListenerForSingleValueEvent(object : ValueEventListener{
            override fun onDataChange(snapshot: DataSnapshot) {
                val children = snapshot.children
                children.forEach {
                    val curCategory = it.getValue(Category::class.java)
                    if (curCategory != null) {
                        categoryList.add(curCategory)
                    }
                }
                for(category in categoryList){
                    resultItemList.add(RecyclerViewItem(category))
                    for(quiz in category.quizzes){
                        resultItemList.add(RecyclerViewItem(quiz))
                    }
                }
                val recyclerView = findViewById<RecyclerView>(R.id.quizRecyclerView)
                val adapter = RecyclerViewAdapter(applicationContext, resultItemList)
                recyclerView.adapter = adapter
            }

            override fun onCancelled(error: DatabaseError) {

            }

        })
    }
}